client/src/main/java/io/pravega/client/stream/impl/StreamSegments.java

```diff
@@ -67,19 +71,28 @@ public Segment getSegmentForKey(double key) {
         return segments.values();
     }
     
-    public StreamSegments withReplacementRange(StreamSegmentsWithPredecessors replacementRanges) {
+    public StreamSegments withReplacementRange(Segment replacedSegment, StreamSegmentsWithPredecessors replacementRanges) {
+        Preconditions.checkState(segments.containsValue(replacedSegment), "Segment to be replaced should be present in the segment list");
         verifyReplacementRange(replacementRanges);
         NavigableMap<Double, Segment> result = new TreeMap<>();
         Map<Long, List<SegmentWithRange>> replacedRanges = replacementRanges.getReplacementRanges();
-        for (Entry<Double, Segment> exitingSegment : segments.entrySet()) {
-            List<SegmentWithRange> replacements = replacedRanges.get(exitingSegment.getValue().getSegmentId());
-            if (replacements == null || replacements.isEmpty()) {
-                result.put(exitingSegment.getKey(), exitingSegment.getValue());
+        Optional<List<SegmentWithRange>> replacements = Optional.ofNullable(replacedRanges.get(replacedSegment.getSegmentId()));
+        Segment lastSegmentValue = null;
+        for (Entry<Double, Segment> exitingEntry : segments.descendingMap().entrySet()) { // iterate from the highest key.
+            final Segment exitingSegment = exitingEntry.getValue();
+            if (exitingSegment.equals(lastSegmentValue)) {
+                // last value was the same segment, it can be consolidated.
+                continue;
+            }
+            if (exitingSegment.equals(replacedSegment)) {
+                // segment needs to be replaced.
+                replacements.ifPresent(segmentWithRanges -> segmentWithRanges.forEach(segmentWithRange ->
+                        result.put(Math.min(segmentWithRange.getHigh(), exitingEntry.getKey()), segmentWithRange.getSegment())));
```

Comment:
I'm not sure what kind of invariant we are trying to satisfy here. It looks like we are trying to guarantee that the we do not add an entry to the mapping with a high that is higher than the high of the existing entry, but why is that correct in the first place?

A related question is the following. It looks like we might end up with multiple entries in the map pointing to the same segment. This is the case so that we guarantee that the we receive segment sealed from each one of the existing segments before writing to a new segment that spans multiple existing ones. But, I'm wondering whether we need to consolidate the mapping once all segment sealed notifications are received.

Reply:
>I'm not sure what kind of invariant we are trying to satisfy here. It looks like we are trying to guarantee that the we do not add an entry to the mapping with a high that is higher than the high of the existing entry, but why is that correct in the first place?

it is possible it is only adding the segment for part of its range. For example if the new segment is a result of a merge and only one of the predecessors has encountered a seal.

>A related question is the following. It looks like we might end up with multiple entries in the map pointing to the same segment. This is the case so that we guarantee that the we receive segment sealed from each one of the existing segments before writing to a new segment that spans multiple existing ones. But, I'm wondering whether we need to consolidate the mapping once all segment sealed notifications are received.

yes, see above.

Reply:
I think the invariant is that we can only add to the result map ranges that fully overlap with the range of the replaced segment, in which case, I wonder whether it should be the min of `segmentWithRange` and `replacedSegment` instead.

In any case, I'd like to come to an agreement on what the precise invariant statement is and to see it expressed in the form of a comment in the code.

Reply:
Right, but this is very low level, I'm after the high-level property that we are trying to satisfy here and it should be expressed in terms of predecessor and successor ranges.

Reply:
```The successor segment(s)'s range should be limited to the replaced segment's range, there by ensuring that newer writes to the successor(s) happen only for the replaced segment's range.```

Does this look any better?